File: jython/exerc_solutions/02_DicAccessFromJava/src/main/java/es/neodoo/pythoncourse/dictaccess/CustomDict.java
package es.neodoo.pythoncourse.dictaccess;

import java.util.*;

public class CustomDict {
	
    public Map data;

    @SuppressWarnings("rawtypes")
	public CustomDict() {
        data = new HashMap();
    }
    
    public void __setitem__(Object key, Object value) {
        data.put(key, value);
    	System.out.println(String.format("Added key %s with value %s", key, value));
    }
    
    public Object __getitem__(Object key) {
    	if (data.containsKey(key)) {
            Object value = data.get(key);
        	System.out.println(String.format("Found key %s  with value %s", key, value));
            return value;
                
        } else {
        	System.out.println(String.format("Key %s not found", key));
        	return null;
        }
    }
    
    public boolean __contains__(Object key) {
    	
        if (data.containsKey(key)) {
        	System.out.println(String.format("Found key %s", key));
            return true;
        } else {
        	System.out.println(String.format("Key %s not found", key));
            return false;
        }
    }
}

File: jython/exercises/02_DicAccessFromJava/src/main/jython/customdict.py
'''
Created on Jul 27, 2011

@author: Luis Perez

Ejercicio creado con el proposito de mostrar la capacidad de emular parcialmente la funcionalidad
de un diccionario python, implementando los metodos necesarios desde Java, y haciendo uso
de ellos desde jython.

Ayuda:
    Aunque java.util.HashMap ya tiene la funcionalidad requerida, puesto que jython lo integra
    por defecto, nuestra clase hara uso de ella para simplificar la implementacion.
    
    Sera necesario implementar los metodos __contains__, __getitem__, y __setitem__.
    Ver (http://docs.python.org/reference/datamodel.html)
'''
import unittest

from es.neodoo.pythoncourse.dictaccess import CustomDict

class Test(unittest.TestCase):
    
    def test_populate_read_dict(self):
        d = CustomDict()
        
        d['key1'] = "value1";
        d[5] = "value2"
        
        self.assertEquals(d['key1'], "value1")
        self.assertEquals(d[5], "value2")
        
    def test_in_dict_syntax(self):
        """testear que la sintaxis "key" in customDict funciona"""
        d = CustomDict()
        
        d['key1'] = "value1";
        d[5] = "value2"
        
        self.assertTrue("key1" in d)
        self.assertFalse(3 in d)
    
    def test_inexistent_key_return_None(self):
        """Testear que pasar una clave que no existe devuelve None"""
        d = CustomDict()
        assert d['nonexistent'] == None
        
if __name__ == "__main__":
    unittest.main()